File: apps/sdk-typescript/src/services/memory.ts
/**
 * `MemoryService` — typed client for the `memory` service.
 *
 * Mirror of `apps/memory/src/http/routes.ts`:
 *   - POST  /v1/memory                  (store)
 *   - POST  /v1/memory/search           (retrieve)
 *   - GET   /v1/memory/:id              (get)
 *   - PATCH /v1/memory/:id              (update)
 *   - POST  /v1/memory/:id/share        (share)
 */

import { request } from '../http.js';

import type { HttpClientOptions } from '../http.js';

export type MemoryType = 'fact' | 'event' | 'preference' | 'relation';
export type Visibility = 'private' | 'operator' | 'shared' | 'public';

export interface Permissions {
  readonly visibility: Visibility;
  readonly sharedWith?: readonly string[];
}

export interface EmbeddingMeta {
  readonly model: string;
  readonly dim: number;
}

export interface StoreRequest {
  readonly ownerDid: string;
  readonly type: MemoryType;
  readonly text: string;
  readonly payload?: Readonly<Record<string, unknown>>;
  readonly permissions: Permissions;
  readonly encryption?: { readonly enabled: boolean };
}

export interface StoreResponse {
  readonly id: string;
  readonly embedding: EmbeddingMeta;
}

export interface SearchFilters {
  readonly type?: MemoryType;
  readonly ownerDid?: string;
  readonly visibility?: Visibility;
}

export interface SearchRequest {
  readonly queryDid: string;
  readonly queryText: string;
  readonly topK?: number;
  readonly filters?: SearchFilters;
}

export interface SearchHit {
  readonly id: string;
  readonly score: number;
  readonly type: MemoryType;
  readonly ownerDid: string;
  readonly snippet: string;
}

export interface SearchResponse {
  readonly hits: readonly SearchHit[];
}

export interface MemoryRecord {
  readonly id: string;
  readonly ownerDid: string;
  readonly type: MemoryType;
  readonly text: string;
  readonly payload: Readonly<Record<string, unknown>>;
  readonly permissions: { readonly visibility: Visibility; readonly sharedWith: readonly string[] };
  readonly encryption: {
    readonly enabled: boolean;
    readonly algorithm: string;
    readonly keyId: string;
  };
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly version: number;
  readonly embedding: EmbeddingMeta;
}

export interface RetrieveRequest {
  readonly id: string;
  readonly callerDid: string;
}

export interface UpdateRequest {
  readonly id: string;
  readonly callerDid: string;
  readonly text?: string;
  readonly payload?: Readonly<Record<string, unknown>>;
}

export interface UpdateResponse {
  readonly id: string;
  readonly version: number;
  readonly embedding: EmbeddingMeta;
}

export interface ShareRequest {
  readonly id: string;
  readonly callerDid: string;
  readonly shareWith: readonly string[];
  readonly expiresAt?: string;
}

export interface ShareResponse {
  readonly id: string;
  readonly sharedWith: readonly string[];
}

export class MemoryService {
  constructor(
    private readonly opts: HttpClientOptions,
    private readonly baseUrl: string,
  ) {}

  /** POST /v1/memory */
  public async store(body: StoreRequest): Promise<StoreResponse> {
    const data = await request<StoreResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/memory',
      body,
    });
    if (data === undefined) {
      throw new Error('memory.store: empty response body');
    }
    return data;
  }

  /** POST /v1/memory/search */
  public async search(body: SearchRequest): Promise<SearchResponse> {
    const data = await request<SearchResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: '/v1/memory/search',
      body,
    });
    if (data === undefined) {
      throw new Error('memory.search: empty response body');
    }
    return data;
  }

  /** GET /v1/memory/:id?callerDid=... */
  public async retrieve({ id, callerDid }: RetrieveRequest): Promise<MemoryRecord> {
    const data = await request<MemoryRecord>(this.opts, {
      method: 'GET',
      baseUrl: this.baseUrl,
      path: `/v1/memory/${encodeURIComponent(id)}`,
      query: { callerDid },
    });
    if (data === undefined) {
      throw new Error('memory.retrieve: empty response body');
    }
    return data;
  }

  /** PATCH /v1/memory/:id */
  public async update({ id, ...rest }: UpdateRequest): Promise<UpdateResponse> {
    const data = await request<UpdateResponse>(this.opts, {
      method: 'PATCH',
      baseUrl: this.baseUrl,
      path: `/v1/memory/${encodeURIComponent(id)}`,
      body: rest,
    });
    if (data === undefined) {
      throw new Error('memory.update: empty response body');
    }
    return data;
  }

  /** POST /v1/memory/:id/share */
  public async share({ id, ...rest }: ShareRequest): Promise<ShareResponse> {
    const data = await request<ShareResponse>(this.opts, {
      method: 'POST',
      baseUrl: this.baseUrl,
      path: `/v1/memory/${encodeURIComponent(id)}/share`,
      body: rest,
    });
    if (data === undefined) {
      throw new Error('memory.share: empty response body');
    }
    return data;
  }
}

File: apps/sdk-python/src/colber_sdk/services/memory.py
"""``MemoryService`` — typed client for the ``memory`` service.

Mirror of ``apps/sdk-typescript/src/services/memory.ts`` and
``apps/memory/src/http/routes.ts``:

- ``POST  /v1/memory``                 (store)
- ``POST  /v1/memory/search``          (search)
- ``GET   /v1/memory/:id``             (retrieve)
- ``PATCH /v1/memory/:id``             (update)
- ``POST  /v1/memory/:id/share``       (share)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal
from urllib.parse import quote

from .._http import HttpClientOptions, RequestParams, request
from ._convert import from_wire, to_wire

MemoryType = Literal["fact", "event", "preference", "relation"]
Visibility = Literal["private", "operator", "shared", "public"]


@dataclass(frozen=True, slots=True)
class EmbeddingMeta:
    model: str
    dim: int


@dataclass(frozen=True, slots=True)
class StoreResponse:
    id: str
    embedding: EmbeddingMeta


@dataclass(frozen=True, slots=True)
class SearchHit:
    id: str
    score: float
    type: str
    owner_did: str
    snippet: str


@dataclass(frozen=True, slots=True)
class SearchResponse:
    hits: list[SearchHit] = field(default_factory=list)


@dataclass(frozen=True, slots=True)
class MemoryRecordPermissions:
    visibility: str
    shared_with: list[str] = field(default_factory=list)


@dataclass(frozen=True, slots=True)
class MemoryRecordEncryption:
    enabled: bool
    algorithm: str
    key_id: str


@dataclass(frozen=True, slots=True)
class MemoryRecord:
    id: str
    owner_did: str
    type: str
    text: str
    payload: dict[str, Any]
    permissions: MemoryRecordPermissions
    encryption: MemoryRecordEncryption
    created_at: str
    updated_at: str
    version: int
    embedding: EmbeddingMeta


@dataclass(frozen=True, slots=True)
class UpdateResponse:
    id: str
    version: int
    embedding: EmbeddingMeta


@dataclass(frozen=True, slots=True)
class ShareResponse:
    id: str
    shared_with: list[str] = field(default_factory=list)


class MemoryService:
    """Typed client for the ``memory`` service."""

    def __init__(self, opts: HttpClientOptions, base_url: str) -> None:
        self._opts = opts
        self._base_url = base_url

    def store(
        self,
        *,
        owner_did: str,
        type: MemoryType,
        text: str,
        permissions: dict[str, Any],
        payload: dict[str, Any] | None = None,
        encryption: dict[str, Any] | None = None,
    ) -> StoreResponse:
        """``POST /v1/memory``."""
        body_in: dict[str, Any] = {
            "owner_did": owner_did,
            "type": type,
            "text": text,
            "permissions": permissions,
        }
        if payload is not None:
            body_in["payload"] = payload
        if encryption is not None:
            body_in["encryption"] = encryption
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/memory",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("memory.store: empty response body")
        return from_wire(StoreResponse, data)

    def search(
        self,
        *,
        query_did: str,
        query_text: str,
        top_k: int | None = None,
        filters: dict[str, Any] | None = None,
    ) -> SearchResponse:
        """``POST /v1/memory/search``."""
        body_in: dict[str, Any] = {"query_did": query_did, "query_text": query_text}
        if top_k is not None:
            body_in["top_k"] = top_k
        if filters is not None:
            body_in["filters"] = filters
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path="/v1/memory/search",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("memory.search: empty response body")
        return from_wire(SearchResponse, data)

    def retrieve(self, *, id: str, caller_did: str) -> MemoryRecord:
        """``GET /v1/memory/:id?callerDid=...``."""
        data = request(
            self._opts,
            RequestParams(
                method="GET",
                base_url=self._base_url,
                path=f"/v1/memory/{quote(id, safe='')}",
                query={"callerDid": caller_did},
            ),
        )
        if data is None:
            raise RuntimeError("memory.retrieve: empty response body")
        return from_wire(MemoryRecord, data)

    def update(
        self,
        *,
        id: str,
        caller_did: str,
        text: str | None = None,
        payload: dict[str, Any] | None = None,
    ) -> UpdateResponse:
        """``PATCH /v1/memory/:id``."""
        body_in: dict[str, Any] = {"caller_did": caller_did}
        if text is not None:
            body_in["text"] = text
        if payload is not None:
            body_in["payload"] = payload
        data = request(
            self._opts,
            RequestParams(
                method="PATCH",
                base_url=self._base_url,
                path=f"/v1/memory/{quote(id, safe='')}",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("memory.update: empty response body")
        return from_wire(UpdateResponse, data)

    def share(
        self,
        *,
        id: str,
        caller_did: str,
        share_with: list[str],
        expires_at: str | None = None,
    ) -> ShareResponse:
        """``POST /v1/memory/:id/share``."""
        body_in: dict[str, Any] = {"caller_did": caller_did, "share_with": share_with}
        if expires_at is not None:
            body_in["expires_at"] = expires_at
        data = request(
            self._opts,
            RequestParams(
                method="POST",
                base_url=self._base_url,
                path=f"/v1/memory/{quote(id, safe='')}/share",
                body=to_wire(body_in),
            ),
        )
        if data is None:
            raise RuntimeError("memory.share: empty response body")
        return from_wire(ShareResponse, data)
